---
title: "Holoruminant MetaG pipeline"
subtitle: "Preprocessing module"
author: "t.b.a."
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: false
    number_sections: true
    theme: cosmo
    df_print: paged
    code_folding: show
---

```{r setup, include=FALSE}
library("GenomicTools")
library("xtable")   # Needed for LaTeX output of the tables
library("viridis")  # Needed for the colouring of the plots
library("rjson")    # Needed for multiqc dgsb etail data
library("DT")
library("kableExtra")
options(scipen=999,
        stringsAsFactors=FALSE)

knitr::opts_chunk$set(echo = FALSE,
                      cache = FALSE,
                      cache.lazy = FALSE,
                      dev = c('png', 'pdf'),
                      fig.align = 'center', fig.height = 5, fig.width = 8.5)

reportColours <- c("#FF8200", "#00B5E2", "#78BE20", "#7F3F98", "#E13C98", "#0033A0", "#54585A")

if(!is.element("snakemake",ls())){
  features_file <- "/scratch/project_2010176/metaG_groupAssembly/config/features.yaml"
  project_folder <- "/scratch/project_2010176/metaG_groupAssembly"
}
```

```{r, include=FALSE}
# Init Step to make sure that the dependencies are loaded
htmltools::tagList(datatable(cars))
#htmltools::tagList(ggplotly(ggplot()))
```

```{r}
sample_sheet <- read.table(file.path(project_folder, "config", "samples.tsv"), header=TRUE) 

nSamples <- nrow(sample_sheet)
```

```{r help functions}
# Define the function that extracts values from the list based on the given field path
extract_values <- function(input_list, field_path) {
  
  # Initialize an empty named vector
  result_vector <- c()
  
  # Loop over each item in the list
  for (item_name in names(input_list)) {
    
    # Construct the full expression to access the desired field
    field_expr <- paste0("input_list[['", item_name, "']]", field_path)
    
    # Evaluate the expression to get the field value
    field_value <- eval(parse(text = field_expr))
    
    # Store the field value in the result vector with the item name
    result_vector[item_name] <- field_value
  }
  
  # Return the named vector
  return(result_vector)
}

number_from_field <- function(vec, x){
  as.numeric(regmatches(vec[grep(x, vec)], regexpr("\\d+", vec[grep(x, vec)])))
}
```

# Overview

The preprocess module runs a couple of tools, mainly:

* fastp	
* bowtie2 
* humann
* kraken2
* metaphlan 
* nonpareil
* phyloflash
* singlem

# FASTP

The summary of the fastp results

```{r import the fastp jason files}
# Get a list of all JSON files in the directory
  json_dir <- file.path(project_folder, "results", "preprocess", "fastp")
  json_files <- list.files(path = json_dir, pattern = "*.json", full.names = TRUE)


# Initialize a list to store the data frames
  json_data_list <- list()

# Loop over each file and read the JSON data
  for (file in json_files) {
  # Read the JSON file into R
    json_data <- fromJSON(file = file)
    
  # Append the data frame to the list
    json_data_list[[file]] <- json_data
  }

# Adjsut the names
  names(json_data_list) <- sapply(strsplit(names(json_data_list), "/fastp/"),"[", 2)
  names(json_data_list) <- gsub("_fastp.json", "", names(json_data_list))
```


```{r}
plot_and_table_fastp_feature <- function(x, prefix1="$summary$before_filtering$", prefix2="$summary$after_filtering$"){
# Get the total input reads
  field_path <- paste0(prefix1,x)  #total_reads"
  result_vector <- extract_values(json_data_list, field_path)
  
  field_path <- paste0(prefix2,x)  #total_reads"
  result_vector_2 <- extract_values(json_data_list, field_path)

# Print the resulting named vector
  barplot(rbind(result_vector, result_vector_2), las=2, col=c(reportColours[1], reportColours[2]), beside=TRUE)  

  tmp <- cbind(names(result_vector),result_vector, result_vector_2)
  if(nrow(tmp)>0){
  rownames(tmp) <- 1:nrow(tmp)
  colnames(tmp) <- c("Sample", paste(x,"(Raw)"), paste(x, "(after filtering)"))
  
  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))
  }  

}
```

## Total reads

```{r}
plot_and_table_fastp_feature("total_reads")
```

## Total bases

```{r}
plot_and_table_fastp_feature("total_bases")
```

## Q20 bases

```{r}
plot_and_table_fastp_feature("q20_bases")
```


## Q30 bases

```{r}
plot_and_table_fastp_feature("q30_bases")
```

## Q20 rate

```{r}
plot_and_table_fastp_feature("q20_rate")
```

## Q30 rate

```{r}
plot_and_table_fastp_feature("q30_rate")
```

## Read1 mean length

```{r}
plot_and_table_fastp_feature("read1_mean_length")
```


## Read2 mean length

```{r}
plot_and_table_fastp_feature("read2_mean_length")
```

## GC content

```{r}
plot_and_table_fastp_feature("gc_content")
```


## Duplication

```{r}
plot_and_table_fastp_one_feature <- function(x, prefix = ""){
# Get the total input reads
  field_path <- paste0(prefix, x)  #total_reads"
  result_vector <- extract_values(json_data_list, field_path)
  
# Print the resulting named vector
  barplot(unlist(result_vector), las=2, col=c(reportColours[1]))

  tmp <- cbind(names(result_vector),result_vector)
  
  if(nrow(tmp)>0){
  rownames(tmp) <- 1:nrow(tmp)
  colnames(tmp) <- c("Sample", x)
  
  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))
  }    
}

```


## Duplication

```{r}
plot_and_table_fastp_one_feature(x="$duplication")
```


## Adapter cutting

### Trimmed reads

```{r}
plot_and_table_fastp_one_feature(x="$adapter_trimmed_reads", "$adapter_cutting")
```

### Trimmed bases

```{r}
plot_and_table_fastp_one_feature(x="$adapter_trimmed_bases", "$adapter_cutting")
```

## Adapter sequences

```{r}

  prefix <- "$adapter_cutting"
  x <- "$read1_adapter_sequence"
  x2 <- "$read2_adapter_sequence"
  field_path <- paste0(prefix, x)
  result_vector <- extract_values(json_data_list, field_path)
  field_path <- paste0(prefix, x2)
  result_vector_2 <- extract_values(json_data_list, field_path)
  

  tmp <- cbind(names(result_vector),result_vector, result_vector_2)
  
  if(nrow(tmp)>0){
  rownames(tmp) <- 1:nrow(tmp)
  colnames(tmp) <- c("Sample", x, x2)
  
  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))
  }   
```

## Adapter counts (Read 1)

```{r}

extract_adapter_counts <- function(json_data_list) {
  adapter_list <- list()

  # Loop through all samples
  for (sample in names(json_data_list)) {
    adapter_counts <- json_data_list[[sample]]$adapter_cutting$read1_adapter_counts
    for (adapter in names(adapter_counts)) {
      if (!is.null(adapter_list[[adapter]])) {
        adapter_list[[adapter]][sample] <- adapter_counts[[adapter]]
      } else {
        # Initialize a named numeric vector
        adapter_list[[adapter]] <- setNames(rep(0, length(json_data_list)), names(json_data_list))
        adapter_list[[adapter]][sample] <- adapter_counts[[adapter]]
      }
    }
  }

  # Convert list to matrix
  adapter_matrix <- do.call(rbind, adapter_list)

  return(adapter_matrix)
}


# Function to create a stacked barplot
plot_adapter_counts <- function(adapter_matrix) {
  barplot(adapter_matrix, beside = FALSE, col = rainbow(nrow(adapter_matrix)), 
          legend.text = rownames(adapter_matrix), las = 2, 
          main = "Adapter Sequences Across Samples", ylab = "Read Counts")
}

# Function to print a table of adapter counts
print_adapter_table <- function(adapter_matrix) {
  cat("Adapter Count Table:\n")
  print(adapter_matrix)
}

# Generate adapter matrix
adapter_matrix <- extract_adapter_counts(json_data_list)

# Print table
tmp <- cbind(rownames(adapter_matrix), adapter_matrix)

  rownames(tmp) <- 1:nrow(tmp)

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))



# Generate stacked barplot
plot_adapter_counts(adapter_matrix)
```



## Adapter counts (Read 2)

```{r}

extract_adapter_counts <- function(json_data_list) {
  adapter_list <- list()

  # Loop through all samples
  for (sample in names(json_data_list)) {
    adapter_counts <- json_data_list[[sample]]$adapter_cutting$read2_adapter_counts
    for (adapter in names(adapter_counts)) {
      if (!is.null(adapter_list[[adapter]])) {
        adapter_list[[adapter]][sample] <- adapter_counts[[adapter]]
      } else {
        # Initialize a named numeric vector
        adapter_list[[adapter]] <- setNames(rep(0, length(json_data_list)), names(json_data_list))
        adapter_list[[adapter]][sample] <- adapter_counts[[adapter]]
      }
    }
  }

  # Convert list to matrix
  adapter_matrix <- do.call(rbind, adapter_list)

  return(adapter_matrix)
}


# Function to create a stacked barplot
plot_adapter_counts <- function(adapter_matrix) {
  barplot(adapter_matrix, beside = FALSE, col = rainbow(nrow(adapter_matrix)), 
          legend.text = rownames(adapter_matrix), las = 2, 
          main = "Adapter Sequences Across Samples", ylab = "Read Counts")
}

# Function to print a table of adapter counts
print_adapter_table <- function(adapter_matrix) {
  cat("Adapter Count Table:\n")
  print(adapter_matrix)
}

# Generate adapter matrix
adapter_matrix <- extract_adapter_counts(json_data_list)

tmp <- cbind(rownames(adapter_matrix), adapter_matrix)

  rownames(tmp) <- 1:nrow(tmp)

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))

# Generate stacked barplot
plot_adapter_counts(adapter_matrix)
```




# Decontamination

## Bowtie2

First, a list of genomes that were used for decontamination

```{r}
# Get all folders as before
all_folders <- list.dirs(file.path(project_folder, "results", "preprocess", "bowtie2"), full.names = FALSE)

# Filter out unwanted entries
filtered_folders <- all_folders[!grepl("^non|^benchmark|^decontaminated_reads|^$", all_folders)]

# Define a function to extract only the aligned reads from the log file
extract_alignment_info <- function(log_file) {
  # Read the content of the log file
  log_content <- readLines(log_file)
  
  # Use regular expressions to find the total number of reads
  total_reads <- as.numeric(gsub(" reads; of these:", "", log_content[grep("reads; of these:", log_content)]))
  
  # Extract the number of reads that aligned concordantly 0 times
  aligned_zero <- as.numeric(strsplit(log_content[grep("aligned concordantly 0 times", log_content)[1]], "\\(")[[1]][1])
  
  # Calculate the number of aligned reads
  aligned_reads <- total_reads - aligned_zero
  
  # Return only aligned reads (non-aligned reads are omitted)
  return(aligned_reads)
}

# Initialize an empty list to store alignment information for each sample
alignment_data <- matrix(0, nrow=nSamples, ncol= length(filtered_folders))

colnames(alignment_data) <- filtered_folders

# Loop through all folders and find all *lib1.log files
folder_running_index <- 1
for (folder in filtered_folders) {
  # List all log files that end with "lib1.log"
  log_files <- list.files(file.path(project_folder, "results", "preprocess", "bowtie2", folder), pattern = "lib1.log$", full.names = TRUE)
  cur_sample_names <- sub(".*/([^/]+)\\.lib1\\.log$", "\\1", log_files)

  rownames(alignment_data) <- cur_sample_names
  
  sample_running_index <- 1
  # Loop through each log file
  for (log_file in log_files) {
    # Extract the alignment information for each log file
    alignment_info <- extract_alignment_info(log_file)
    
    # Store the aligned reads in the list (using sample name as the key)
    alignment_data[sample_running_index, folder_running_index] <- alignment_info
    
    sample_running_index <- sample_running_index + 1
  }
  
  folder_running_index <- folder_running_index + 1
}

tmp <- cbind(rownames(alignment_data), alignment_data)

  rownames(tmp) <- 1:nrow(tmp)

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))



```



## Quantification
t.b.a

# Diamond



```{r}
diamond_groups <- list.dirs(file.path(project_folder, "results", "preprocess", "diamond"), recursive = FALSE, full.names = FALSE)

result_matrix_diamond <- matrix(nrow=nSamples, ncol=3*length(diamond_groups))
colnames(result_matrix_diamond) <- 1:(3*length(diamond_groups))
rownames(result_matrix_diamond) <- paste(sample_sheet$sample_id, sample_sheet$library_id, sep="_")

diamond_run <- 1
sample_run <- 1

for(diamond_run in 1:length(diamond_groups)){
 
  for(sample_run in 1:nSamples){
    
    tmp <- readLines(file.path(file.path(project_folder, "results", "preprocess", "diamond",
                                         paste0(diamond_groups[diamond_run],"_", sample_sheet$sample_id[sample_run], "_", sample_sheet$library_id[sample_run], ".log" ))))
    
    total_aligned <- as.numeric(regmatches(tmp[grep("Reported", tmp)], regexpr("\\d+", tmp[grep("Reported", tmp)])))
    queries_aligned <- as.numeric(regmatches(tmp[grep("queries aligned", tmp)], regexpr("\\d+", tmp[grep("queries aligned", tmp)])))
    
    result_matrix_diamond[sample_run, (diamond_run-1)*3 + 1 ] <- queries_aligned
    result_matrix_diamond[sample_run, (diamond_run-1)*3 + 2 ] <- total_aligned
    result_matrix_diamond[sample_run, (diamond_run-1)*3 + 3 ] <- round(queries_aligned/total_aligned,5)
  }
  
  colnames(result_matrix_diamond)[(diamond_run-1)*3 + 1] <- paste0("Queries_aligned_", diamond_groups[diamond_run])
  colnames(result_matrix_diamond)[(diamond_run-1)*3 + 2] <- paste0("Total_aligned_", diamond_groups[diamond_run])
  colnames(result_matrix_diamond)[(diamond_run-1)*3 + 3] <- paste0("Query_ratio_", diamond_groups[diamond_run])
}


  tmp <- result_matrix_diamond

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))

```

# Humann

```{r}
result_matrix_humann <- matrix(nrow=nSamples, ncol=5)
colnames(result_matrix_humann) <- 1:5
rownames(result_matrix_humann) <- paste(sample_sheet$sample_id, sample_sheet$library_id, sep="_")

sample_run <- 1

for(sample_run in 1:nSamples){
    
  tmp <- readLines(file.path(file.path(project_folder, "results", "preprocess", "humann", "log",
                                       paste0(sample_sheet$sample_id[sample_run], ".", sample_sheet$library_id[sample_run], ".log" ))))
    
    species_prescreen <- number_from_field(tmp, "from prescreen") 
    gene_families_from_nucleotide_alignment <- number_from_field(tmp, "gene families from nucleotide alignment")
    unaligned_reads_after_nucleotide_alignment <- number_from_field(tmp, "Unaligned reads after nucleotide alignment")
    total_gene_families_after_translated_alignment <- number_from_field(tmp, "Total gene families after translated alignment")
    unaligned_reads_after_translated_alignment <- number_from_field(tmp, "Unaligned reads after translated alignment")
    
    result_matrix_humann[sample_run, 1 ] <- species_prescreen
    result_matrix_humann[sample_run, 2 ] <- gene_families_from_nucleotide_alignment
    result_matrix_humann[sample_run, 3 ] <- unaligned_reads_after_nucleotide_alignment
    result_matrix_humann[sample_run, 4 ] <- total_gene_families_after_translated_alignment
    result_matrix_humann[sample_run, 5 ] <- unaligned_reads_after_translated_alignment
  }
  
  colnames(result_matrix_humann) <- c("species_prescreen",
                                      "gene_families_from_nucleotide_alignment",
                                      "unaligned_reads_after_nucleotide_alignment",
                                      "total_gene_families_after_translated_alignment",
                                      "unaligned_reads_after_translated_alignment")


  tmp <- result_matrix_humann

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))

```

# Kraken2

```{r}
kraken_groups <- list.dirs(file.path(project_folder, "results", "preprocess", "kraken2"), recursive = FALSE, full.names = FALSE)
kraken_groups <- setdiff(kraken_groups, "benchmark")

result_matrix_kraken <- matrix(nrow=nSamples, ncol=5*length(kraken_groups))
colnames(result_matrix_kraken) <- 1:(5*length(kraken_groups))
rownames(result_matrix_kraken) <- paste(sample_sheet$sample_id, sample_sheet$library_id, sep="_")

kraken_run <- 1
sample_run <- 1


for(kraken_run in 1:length(kraken_groups)){
  
  for(sample_run in 1:nSamples){
    
  tmp <- readLines(file.path(file.path(project_folder, "results", "preprocess", "kraken2", kraken_groups[kraken_run],
                                       paste0(sample_sheet$sample_id[sample_run], ".", sample_sheet$library_id[sample_run], ".log" ))))
    
    number_sequences <- number_from_field(tmp[2], "")
    classified_sequences <- number_from_field(tmp[3], "")
    unclassified_sequences <- number_from_field(tmp[4], "")
    
    result_matrix_kraken[sample_run, (kraken_run-1) * 5 + 1 ] <- number_sequences
    result_matrix_kraken[sample_run, (kraken_run-1) * 5 + 2 ] <- classified_sequences
    result_matrix_kraken[sample_run, (kraken_run-1) * 5 + 3 ] <- round(classified_sequences / number_sequences, 4)
    result_matrix_kraken[sample_run, (kraken_run-1) * 5 + 4 ] <- unclassified_sequences
    result_matrix_kraken[sample_run, (kraken_run-1) * 5 + 5 ] <- round(unclassified_sequences / number_sequences, 4)
  }
  
    colnames(result_matrix_kraken)[(kraken_run-1) * 5 + 1] <- "Number_sequences"
    colnames(result_matrix_kraken)[(kraken_run-1) * 5 + 2] <- "Classified_sequences"
    colnames(result_matrix_kraken)[(kraken_run-1) * 5 + 3] <- "Classified_ratio"
    colnames(result_matrix_kraken)[(kraken_run-1) * 5 + 4] <- "Unclassified_sequences"
    colnames(result_matrix_kraken)[(kraken_run-1) * 5 + 5] <- "Unclassified_ratio"

  

}
  tmp <- result_matrix_kraken

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))

```

# Metaphlan

```{r}
result_matrix_metaphlan <- matrix(nrow=nSamples, ncol=2)
colnames(result_matrix_metaphlan) <- 1:2
rownames(result_matrix_metaphlan) <- paste(sample_sheet$sample_id, sample_sheet$library_id, sep="_")

sample_run <- 1

for(sample_run in 1:nSamples){
    
  tmp <- readLines(file.path(file.path(project_folder, "results", "preprocess", "metaphlan", "profiled",
                                       paste0(sample_sheet$sample_id[sample_run], ".", sample_sheet$library_id[sample_run], ".txt" ))))
    
    processed_reads <- number_from_field(tmp, "reads processed") 
    classifications <- length(tmp) - sum(grepl("^#", tmp))
    
    result_matrix_metaphlan[sample_run, 1 ] <- processed_reads
    result_matrix_metaphlan[sample_run, 2 ] <- classifications
  }
  
  colnames(result_matrix_metaphlan) <- c("processed_reads",
                                      "classifications")


  tmp <- result_matrix_metaphlan

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))

```

# Nonpareil

```{r}

result_matrix_nonpareil <- read.table(file.path(project_folder, "results", "preprocess", "nonpareil", "nonpareil.tsv"), sep="\t", header=TRUE)

  tmp <- result_matrix_nonpareil

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))

```

# Phyloflash
```{r}

result_matrix_phyloflash <- NULL

for(sample_run in 1:nrow(sample_sheet)){
  tar_file <- paste0("/scratch/project_2010176/metaG_groupAssembly/results/preprocess/phyloflash/",
                     sample_sheet$sample_id[sample_run],"_",
                     sample_sheet$library_id[sample_run],".phyloFlash.tar.gz")
  
  csv_file <- paste0(sample_sheet$sample_id[sample_run], "_",
                     sample_sheet$library_id[sample_run], ".phyloFlash.report.csv")  # Name of the file inside the archive
  
  # List files in the archive to verify the correct path
  files_in_tar <- untar(tar_file, list = TRUE)

  # Read the CSV directly from the tar.gz without extracting
  csv_data <- read.table(pipe(paste("tar -xOzf", tar_file, csv_file)), stringsAsFactors = FALSE, sep=",", header=FALSE)
  
   # Rename columns to avoid conflicts (except the first column)
  colnames(csv_data) <- c("ID", paste0(sample_sheet$sample_id[sample_run], "_", sample_sheet$library_id[sample_run]))

  # Merge with existing data
  if (is.null(result_matrix_phyloflash)) {
    result_matrix_phyloflash <- csv_data  # First file initializes the merged dataset
  } else {
    result_matrix_phyloflash <- merge(result_matrix_phyloflash, csv_data, by="ID", all=TRUE)
  }
}


  tmp <- result_matrix_phyloflash
  
  remove_rows <- c("CPUs", "cwd", "database path", "forward read file", "program command", "reverse read file")  # List of row names to remove
  tmp <- tmp[!tmp[,1] %in% remove_rows,]

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))
```


# SingleM

```{r}
result_matrix_singlem <- matrix(nrow=nSamples, ncol=17)
colnames(result_matrix_singlem) <- 1:17
rownames(result_matrix_singlem) <- paste(sample_sheet$sample_id, sample_sheet$library_id, sep="_")

sample_run <- 1

for(sample_run in 1:nSamples){
    
  tmp <- readLines(file.path(file.path(project_folder, "results", "preprocess", "singlem", "pipe",
                                       paste0(sample_sheet$sample_id[sample_run], ".", sample_sheet$library_id[sample_run], ".log" ))))
  
  # First remove the unnecessary part of the log file
  tmp <- tmp[-(1:grep("INFO: Finished$", tmp))]
  tmp <- sapply(strsplit(tmp, "INFO: "), "[", 2)
  
    assigned_out <- number_from_field(tmp, "Found")[1] 
    total_out_coverage_query <- number_from_field(tmp, "Total OTU coverage by query")[1]
    total_out_coverage_diamond <- number_from_field(tmp, "Total OTU coverage by diamond")[1]
    
    # Filter out only the lines containing taxonomy data (those that contain a tab)
    filtered_tmp <- tmp[grepl("\t", tmp)]

    # Split each entry into parts using tab as separator
    split_data <- strsplit(filtered_tmp, "\t")

   # Extract the second column (percentage values) and remove "%" sign
    percentages <- sapply(split_data, function(x) gsub("%", "", x[2]))
    taxons <- sapply(split_data, function(x) gsub(" taxons", "", x[3]))
    
    result_matrix_singlem[sample_run, 1 ] <- assigned_out
    result_matrix_singlem[sample_run, 2 ] <- total_out_coverage_query
    result_matrix_singlem[sample_run, 3 ] <- total_out_coverage_diamond
    # Store values in the result matrix
    result_matrix_singlem[sample_run, 4 ] <- percentages[1]  # Kingdom
    result_matrix_singlem[sample_run, 5 ] <- taxons[1]
    
    result_matrix_singlem[sample_run, 6 ] <- percentages[2]  # Phylum
    result_matrix_singlem[sample_run, 7 ] <- taxons[2]
    
    result_matrix_singlem[sample_run, 8 ] <- percentages[3]  # Class
    result_matrix_singlem[sample_run, 9 ] <- taxons[3]
    
    result_matrix_singlem[sample_run, 10 ] <- percentages[4] # Order
    result_matrix_singlem[sample_run, 11 ] <- taxons[4]
    
    result_matrix_singlem[sample_run, 12 ] <- percentages[5] # Family
    result_matrix_singlem[sample_run, 13 ] <- taxons[5]
    
    result_matrix_singlem[sample_run, 14 ] <- percentages[6] # Genus
    result_matrix_singlem[sample_run, 15 ] <- taxons[6]
    
    result_matrix_singlem[sample_run, 16 ] <- percentages[7] # Species
    result_matrix_singlem[sample_run, 17 ] <- taxons[7]
  
  }
  
# Assign column names
colnames(result_matrix_singlem) <- c("assigned_out", "total_out_coverage_query", "total_out_coverage_diamond",
                                     "kingdom_%", "kingdom_taxons", 
                                     "phylum_%", "phylum_taxons",
                                     "class_%", "class_taxons",
                                     "order_%", "order_taxons",
                                     "family_%", "family_taxons",
                                     "genus_%", "genus_taxons",
                                     "species_%", "species_taxons")

  tmp <- result_matrix_singlem

  datatable(tmp, extensions = 'Buttons',
              options = list(dom = 'Blfrtip',
                             buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                             lengthMenu = list(c(10,25,50,-1),
                                               c(10,25,50,"All"))))

```
