<a name="toc"></a>
# Table of content 
- The [`reads`](#reads) module
- The [`reference`](#reference) module
- The [`preprocess`](#preprocess) module
- The [`read_annotate`](#read_annotate) module
- The [`assemble`](#assemble) module
- The [`contig_annotate`](#contig_annotate) module
- The [`mag_annotate`](#mag_annotate) module
- The [`quantify`](#quantify) module

# Usage

The pipeline can run the entire workflow at once. However, normally it is recommended to run different
modules from the pipeline separated to get better control over the results and also to be able to react
quicker to possible errors.

In the following it is assumed that the pipeline runs on a server that utilizes SLURM and Singularity/Apptainer.
Other setups are also supported, but no default configuration is currently provided. In case you have a
different setup and want to contribute a profile/configuration, please reach out.

For testing and developing, you can add to every command e.g. the option `-np` for a dry-run that prints
the used commands.

The different module have also individual reports that can be generated by adding `report_` in front
of the module name, when a module is called. However, the reports are currently under developments and do
not produce any reasonable output and might crash even.

Back to [Table of contents](#toc)

<a name="reads"></a>
## 'reads'-module
Here are some basic steps for the reads are performed.

Usage:

```
bash run_Pipeline-Holoruminant-meta.sh reads

# Generate the module report
bash run_Pipeline-Holoruminant-meta.sh report_reads
```

For testing, please check first the dry-run with commands printed by running the command like this

```
bash run_Pipeline-Holoruminant-meta.sh reads -np
```

For all other modules this works in a similar fashion, just add the `-np`-option for testing and
`report_` to the module to generate a report for the module (not implemented yet for all modules).

Back to [Table of contents](#toc)

<a name="reference"></a>
## 'reference'-module

The reference host genomes are recompressed in this module.

Usage:

```
bash run_Pipeline-Holoruminant-meta.sh reference
```

Back to [Table of contents](#toc)

<a name="preprocess"></a>
## 'preprocess'-module
This module covers the quality control and trimming of the reads

Usage:

```
bash run_Pipeline-Holoruminant-meta.sh preprocess

# Generate the module report
bash run_Pipeline-Holoruminant-meta.sh report_preprocess
```

Back to [Table of contents](#toc)

<a name="read_annotate"></a>
## 'read_annotate'-module

The command `read_annotate` triggers the analysis of the read-based annotation.

Usage:

```
bash run_Pipeline-Holoruminant-meta.sh read_annotate
```

Instead of the entire module, individual tools can also be called by running

```
bash run_Pipeline-Holoruminant-meta.sh read_annotate__diamond
bash run_Pipeline-Holoruminant-meta.sh read_annotate__humann
bash run_Pipeline-Holoruminant-meta.sh read_annotate__kraken2
bash run_Pipeline-Holoruminant-meta.sh read_annotate__krona
bash run_Pipeline-Holoruminant-meta.sh read_annotate__metaphlan
bash run_Pipeline-Holoruminant-meta.sh read_annotate__nonpareil
bash run_Pipeline-Holoruminant-meta.sh read_annotate__phyloflash
bash run_Pipeline-Holoruminant-meta.sh read_annotate__singlem
bash run_Pipeline-Holoruminant-meta.sh read_annotate__sylph
```

The underlying syntax (here and for all other modules) is that an individual tool is always
called by module name, followed by two underscores and the tool name. Please bear in mind that
the pipeline automatically runs the required pre-steps. This syntax hold through-out the entire
pipeline and can be used to run only subparts of the pipeline.

Back to [Table of contents](#toc)

<a name="assemble"></a>
## 'assemble'-module
This module runs all the assembly related tasks, like creating the metagenome and then the
binning and combination of different binners.

Usage (full module):

```
bash run_Pipeline-Holoruminant-meta.sh assemble
```

Usage (individual tools):

```
bash run_Pipeline-Holoruminant-meta.sh assemble__concoct
bash run_Pipeline-Holoruminant-meta.sh assemble__drep
bash run_Pipeline-Holoruminant-meta.sh assemble__magscot
bash run_Pipeline-Holoruminant-meta.sh assemble__maxbin2
bash run_Pipeline-Holoruminant-meta.sh assemble__megahit
bash run_Pipeline-Holoruminant-meta.sh assemble__metabat2
bash run_Pipeline-Holoruminant-meta.sh assemble__metaspades
```

Back to [Table of contents](#toc)

<a name="contig_annotate"></a>
## 'contig_annotate'-module

This module is running the contig based analysis from beginning to end, meaning it first predicts
genes on the metagenome assembly, annotates the predicted genes with eggnog, aligns the quality
filtered reads to the assembly with bowtie and then quantifies it with coverm.

Usage:

```
bash run_Pipeline-Holoruminant-meta.sh contig_annotate
```

Usage for individual tools:

```
bash run_Pipeline-Holoruminant-meta.sh contig_annotate__camper
bash run_Pipeline-Holoruminant-meta.sh contig_annotate__diamond
bash run_Pipeline-Holoruminant-meta.sh contig_annotate__eggnog
bash run_Pipeline-Holoruminant-meta.sh contig_annotate__eggnog7
bash run_Pipeline-Holoruminant-meta.sh contig_annotate__featurecounts
bash run_Pipeline-Holoruminant-meta.sh contig_annotate__hmmer
bash run_Pipeline-Holoruminant-meta.sh contig_annotate__prodigal
```

Back to [Table of contents](#toc)

<a name="mag_annotate"></a>
## 'mag_annotate'-module


This module is the MAG annotation workhorse, it aligns the mags against various databases and
generates the annotation for them.

Usage:

```
bash run_Pipeline-Holoruminant-meta.sh mag_annotate
```

Individual tools can be accessed like:

```
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__bakta
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__camper
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__checkm2
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__dram
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__dram_mag
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__eggnog
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__gtdbtk
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__phylophlan
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__proteinortho
bash run_Pipeline-Holoruminant-meta.sh mag_annotate__quast
```

Back to [Table of contents](#toc)

<a name="quantify"></a>
## 'quantify'-module

After creating the assemblies, this module does the mapping of the reads and generates the
quantification tables for the samples.

Usage:

```
bash run_Pipeline-Holoruminant-meta.sh quantify
```